File: Java & Python/arrays/next_permutation.java
public class Solution {
	public void nextPermutation(ArrayList<Integer> A) {
		boolean status;
		status = nextPerm(A);
		if (!status)
			Collections.sort(A);
		return;
	}

	public boolean nextPerm(ArrayList<Integer> A) {
		int i = 0; 
		int n = A.size();

		for (i = n-2; i >= 0; --i) 
			if (A.get(i) < A.get(i+1))
				break;
		
		if (i == -1)	return false;

		int j = n-1;
		for(; j >= i; --j) 
			if (A.get(j) > A.get(i))
				break;

		swap (A, i, j);
		++i;

		int steps = (n-i+1)/2;
		for (int k = 0; k < steps; ++k)
			swap(A, i+k, n-k-1);

		return true;
	}

	public void swap(ArrayList<Integer> A, int i, int j) {
		int temp = A.get(i);
		A.set(i, A.get(j));
		A.set(j, temp);
	}
}

/*
	 1, 2, 3, 4, 5
	---------------
	i = 3, j = 4
	4 >= 3

	1, 2, 3, 5, 4
	i = 4
	steps = (5-4+1)/2 = 1
	k = 0
	swap(4, 5-0-1)
 */

File: Java & Python/arrays/next_permutation.py
class Solution:
    # @param A : list of integers
    # @return the same list of integer after modification
    def nextPermutation(self, A):
    	firstChar = -1
    	n = len(A)
    	for i in xrange(n-2, -1, -1):
    		if A[i] < A[i+1]:
    			firstChar = i
    			break
    	if firstChar == -1:
    		return sorted(A)
    	
    	secondChar = left = firstChar + 1
    	right = n - 1

    	for i in xrange(left+1, right+1):
    		if A[firstChar] < A[i] and A[i] < A[secondChar]:
    			secondChar = i

    	A[firstChar], A[secondChar] = A[secondChar], A[firstChar]	# swap

    	A = A[:firstChar + 1] + sorted(A[firstChar+1:])
    	
    	return A

"""
Implement the next permutation, which rearranges numbers into the numerically next greater permutation of numbers.

If such arrangement is not possible, it must be rearranged as the lowest possible order ie, sorted in an ascending order.

The replacement must be in-place, do not allocate extra memory.

Examples:

1,2,3 → 1,3,2

3,2,1 → 1,2,3

1,1,5 → 1,5,1

20, 50, 113 → 20, 113, 50
-------------------------------
0 1 2 3 4
---------
1 2 3 4 5

firstChar = 3
secondChar = 4
left = 4
right = 4

1 2 3 5 4
return A[0:4] - A[4:5]
"""